File: Example/Realm/SyncKitRealmExample/QSCompanyTableViewController.m
//
//  QSCompanyTableViewController.m
//  SyncKitRealm
//
//  Created by Manuel Entrena on 04/05/2017.
//  Copyright © 2017 Colourbox. All rights reserved.
//

#import "QSCompanyTableViewController.h"
#import "QSEmployeeTableViewController.h"

#import <Realm/Realm.h>

#import "QSCompanyTableViewCell.h"
#import "QSCompany.h"
@import SyncKit;

@interface QSCompanyTableViewController () <UICloudSharingControllerDelegate>

@property (nonatomic, strong) RLMResults<QSCompany *> *companies;

@property (nonatomic, strong) RLMNotificationToken *notificationToken;

@property (nonatomic, weak) IBOutlet UIButton *syncButton;
@property (nonatomic, weak) IBOutlet UIActivityIndicatorView *indicatorView;

@property (nonatomic, strong) QSCompany *sharingCompany;

@end

@implementation QSCompanyTableViewController

- (void)awakeFromNib
{
    [super awakeFromNib];
}

- (void)viewDidLoad {
    [super viewDidLoad];
    // Do any additional setup after loading the view, typically from a nib.
    [self setupCompanies];
}

- (void)setupCompanies
{
    self.companies = [[QSCompany allObjectsInRealm:self.realm] sortedResultsUsingKeyPath:@"sortIndex" ascending:YES];
    
    __weak QSCompanyTableViewController *weakSelf = self;
    self.notificationToken = [_companies addNotificationBlock:^(RLMResults<QSCompany *> *results, RLMCollectionChange *changes, NSError *error) {
        if (error) {
            NSLog(@"Failed to open Realm on background worker: %@", error);
            return;
        }
        
        UITableView *tableView = weakSelf.tableView;
        // Initial run of the query will pass nil for the change information
        if (!changes) {
            [tableView reloadData];
            return;
        }
        
        // Query results have changed, so apply them to the UITableView
        [tableView beginUpdates];
        [tableView deleteRowsAtIndexPaths:[changes deletionsInSection:0]
                         withRowAnimation:UITableViewRowAnimationAutomatic];
        [tableView insertRowsAtIndexPaths:[changes insertionsInSection:0]
                         withRowAnimation:UITableViewRowAnimationAutomatic];
        [tableView reloadRowsAtIndexPaths:[changes modificationsInSection:0]
                         withRowAnimation:UITableViewRowAnimationAutomatic];
        [tableView endUpdates];
    }];
}

- (void)viewWillAppear:(BOOL)animated {
    self.clearsSelectionOnViewWillAppear = self.splitViewController.isCollapsed;
    [super viewWillAppear:animated];
}


- (void)didReceiveMemoryWarning {
    [super didReceiveMemoryWarning];
    // Dispose of any resources that can be recreated.
}


- (IBAction)insertNewCompany:(id)sender
{
    UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"New company" message:nil preferredStyle:UIAlertControllerStyleAlert];
    [alertController addTextFieldWithConfigurationHandler:^(UITextField * _Nonnull textField) {
        textField.placeholder = @"Enter company's name";
    }];
    [alertController addAction:[UIAlertAction actionWithTitle:@"Add" style:UIAlertActionStyleDefault handler:^(UIAlertAction * _Nonnull action) {
        [self createCompanyWithName:alertController.textFields.firstObject.text];
    }]];
    [self presentViewController:alertController animated:YES completion:nil];
}

- (void)createCompanyWithName:(NSString *)name
{
    QSCompany *newCompany = [[QSCompany alloc] init];
    newCompany.name = name;
    newCompany.identifier = [[NSUUID UUID] UUIDString];
    newCompany.sortIndex = @(self.companies.count);
    
    [self.realm beginWriteTransaction];
    [self.realm addObject:newCompany];
    [self.realm commitWriteTransaction];
}

- (IBAction)didTapSynchronize:(id)sender
{
    [self synchronizeWithCompletion:nil];
}

#pragma mark - Segues

- (void)prepareForSegue:(UIStoryboardSegue *)segue sender:(id)sender {
    if ([segue.identifier isEqualToString:@"showEmployees"]) {
        NSIndexPath *indexPath = [self.tableView indexPathForSelectedRow];
        QSCompany *company = self.companies[indexPath.row];
        QSEmployeeTableViewController *controller = (QSEmployeeTableViewController *)segue.destinationViewController;
        controller.realm = self.realm;
        controller.company = company;
        controller.canWrite = YES;
        controller.navigationItem.leftBarButtonItem = self.splitViewController.displayModeButtonItem;
        controller.navigationItem.leftItemsSupplementBackButton = YES;
    }
}


#pragma mark - Table View

- (NSInteger)numberOfSectionsInTableView:(UITableView *)tableView {
    return 1;
}


- (NSInteger)tableView:(UITableView *)tableView numberOfRowsInSection:(NSInteger)section {
    return self.companies.count;
}


- (UITableViewCell *)tableView:(UITableView *)tableView cellForRowAtIndexPath:(NSIndexPath *)indexPath {
    QSCompanyTableViewCell *cell = [tableView dequeueReusableCellWithIdentifier:@"cell" forIndexPath:indexPath];

    QSCompany *company = self.companies[indexPath.row];
    
    cell.nameLabel.text = company.name;
    
    CKShare *share = [self.synchronizer shareFor:company];
    if (share) {
        [cell.sharingButton setTitle:@"Sharing" forState:UIControlStateNormal];
    } else {
        [cell.sharingButton setTitle:@"Share" forState:UIControlStateNormal];
    }
    
    __weak QSCompanyTableViewController *weakSelf = self;
    
    cell.shareButtonAction = ^{
        [weakSelf shareCompany:company];
    };
    
    return cell;
}


- (BOOL)tableView:(UITableView *)tableView canEditRowAtIndexPath:(NSIndexPath *)indexPath {
    // Return NO if you do not want the specified item to be editable.
    return YES;
}


- (void)tableView:(UITableView *)tableView commitEditingStyle:(UITableViewCellEditingStyle)editingStyle forRowAtIndexPath:(NSIndexPath *)indexPath {
    if (editingStyle == UITableViewCellEditingStyleDelete) {
        QSCompany *company = [self.companies objectAtIndex:indexPath.row];
        [self.realm transactionWithBlock:^{
            for (QSEmployee *employee in company.employees) {
                [self.realm deleteObject:employee];
            }
            [self.realm deleteObject:company];
        }];
    }
}

- (void)tableView:(UITableView *)tableView didSelectRowAtIndexPath:(NSIndexPath *)indexPath
{
    QSCompany *company = [self.companies objectAtIndex:indexPath.row];
    [self performSegueWithIdentifier:@"showEmployees" sender:company];
    
    [self.tableView deselectRowAtIndexPath:indexPath animated:YES];
}

#pragma mark - Sync

- (void)synchronizeWithCompletion:(void(^)(NSError *error))completion
{
    [self showLoading:YES];
    __weak QSCompanyTableViewController *weakSelf = self;
    [self.synchronizer synchronizeWithCompletion:^(NSError *error) {
        
        [weakSelf showLoading:NO];
        
        if (error) {
            
            if (error.code == CKErrorChangeTokenExpired) {
                [self.appDelegate didGetChangeTokenExpiredError];
            } else {
                UIAlertController *alertController = [UIAlertController alertControllerWithTitle:@"Error" message:[NSString stringWithFormat:@"Error: %@", error] preferredStyle:UIAlertControllerStyleAlert];
                
                [alertController addAction:[UIAlertAction actionWithTitle:@"OK" style:UIAlertActionStyleDefault handler:nil]];
                
                [weakSelf presentViewController:alertController animated:YES completion:nil];
            }
        } else {
            [self.tableView reloadData];
            CKRecordZoneID *zoneID = weakSelf.synchronizer.modelAdapters.firstObject.recordZoneID;
            if (zoneID) {
                [weakSelf.synchronizer subscribeForChangesIn:zoneID completion:^(NSError *error) {
                    if (error) {
                        NSLog(@"Failed to subscribe with error: %@", error);
                    } else {
                        NSLog(@"Subscribed for notifications");
                    }
                }];
            }
        }
        if (completion) {
            completion(error);
        }
    }];
}

#pragma mark - Sharing

- (void)shareCompany:(QSCompany *)company
{
    self.sharingCompany = company;
    
    __weak QSCompanyTableViewController *weakSelf = self;
    
    [self synchronizeWithCompletion:^(NSError *error) {
        
        if (!error) {
            UICloudSharingController *sharingController;
            CKShare *share = [weakSelf.synchronizer shareFor:company];
            CKContainer *container = [CKContainer containerWithIdentifier:weakSelf.synchronizer.containerIdentifier];
            if (share) {
                share[CKShareTitleKey] = company.name;
                sharingController = [[UICloudSharingController alloc] initWithShare:share container:container];
            } else {
                sharingController = [[UICloudSharingController alloc] initWithPreparationHandler:^(UICloudSharingController * _Nonnull controller, void (^ _Nonnull preparationCompletionHandler)(CKShare * _Nullable, CKContainer * _Nullable, NSError * _Nullable)) {
                    
                    [self.synchronizer shareWithObject:company publicPermission:CKShareParticipantPermissionReadOnly participants:@[] completion:^(CKShare *share, NSError *error) {
                        
                        share[CKShareTitleKey] = company.name;
                        preparationCompletionHandler(share, container, error);
                    }];
                }];
            }
            
            sharingController.availablePermissions = UICloudSharingPermissionAllowPublic | UICloudSharingPermissionAllowReadOnly | UICloudSharingPermissionAllowReadWrite;
            sharingController.delegate = self;
            [weakSelf presentViewController:sharingController animated:YES completion:nil];
        }
    }];
}

#pragma mark - UICloudSharingControllerDelegate

- (NSString *)itemTitleForCloudSharingController:(UICloudSharingController *)csc
{
    return self.sharingCompany.name;
}

- (void)cloudSharingController:(UICloudSharingController *)csc failedToSaveShareWithError:(NSError *)error
{
    NSLog(@"Error saving: %@", error);
}

- (void)cloudSharingControllerDidSaveShare:(UICloudSharingController *)csc
{
    [self.synchronizer saveShare:csc.share for:self.sharingCompany];
    [self.tableView reloadData];
}

- (void)cloudSharingControllerDidStopSharing:(UICloudSharingController *)csc
{
    [self.synchronizer deleteShareFor:self.sharingCompany];
    [self.tableView reloadData];
}

#pragma mark - Loading

- (void)showLoading:(BOOL)loading
{
    if (loading) {
        self.syncButton.hidden = YES;
        [self.indicatorView startAnimating];
    } else {
        self.syncButton.hidden = NO;
        [self.indicatorView stopAnimating];
    }
}

#pragma mark - Resetting

- (void)stopUsingRealmObjects
{
    [self.notificationToken invalidate];
    self.sharingCompany = nil;
    self.companies = nil;
    [self.tableView reloadData];
}

@end

File: Example/RealmSwift/SyncKitRealmSwift/QSCompanyTableViewController.swift
//
//  QSCompanyTableViewController.swift
//  SyncKitRealmSwiftExample
//
//  Created by Manuel Entrena on 01/09/2017.
//  Copyright © 2017 Manuel Entrena. All rights reserved.
//

import UIKit
import RealmSwift
import SyncKit

class QSCompanyTableViewController: UITableViewController, UICloudSharingControllerDelegate {
    
    var realm: Realm!
    var synchronizer: QSCloudKitSynchronizer!
    weak var appDelegate: AppDelegate!
    
    var notificationToken: NotificationToken!
    
    @IBOutlet weak var syncButton: UIButton?
    @IBOutlet weak var indicatorView: UIActivityIndicatorView?
    
    var companies: Results<QSCompany>?
    
    var sharingCompany: QSCompany?
    
    override func viewDidLoad() {
        super.viewDidLoad()
        
        setupCompanies()
    }
    
    func setupCompanies() {
        
        companies = realm.objects(QSCompany.self).sorted(byKeyPath: "sortIndex")
        
        notificationToken = companies?.observe({ [weak self] (change) in
            switch change {
            case .error(_):
                
                print("Realm error")
                break
            case .update(_, let deletions, let insertions, let modifications):
                
                self?.tableView.beginUpdates()
                self?.tableView.deleteRows(at: deletions.map { IndexPath(row: $0, section: 0) }, with: .automatic)
                self?.tableView.insertRows(at: insertions.map { IndexPath(row: $0, section: 0) }, with: .automatic)
                self?.tableView.reloadRows(at: modifications.map { IndexPath(row: $0, section: 0) }, with: .automatic)
                self?.tableView.endUpdates()
                
            default:
                
                self?.tableView.reloadData()
            }
        })
    }

    override func viewWillAppear(_ animated: Bool) {
        super.viewWillAppear(animated)
        
        clearsSelectionOnViewWillAppear = splitViewController?.isCollapsed ?? true
    }
    
    @IBAction func insertNewCompany() {
        
        let alertController = UIAlertController(title: "New company", message: nil, preferredStyle: .alert)
        alertController.addTextField { (textField) in
            textField.placeholder = "Enter company's name"
        }
        alertController.addAction(UIAlertAction(title: "Add", style: .default, handler: { (action) in
            
            self.createCompany(name: alertController.textFields!.first!.text!)
        }))
        
        present(alertController, animated: true, completion: nil)
    }
    
    func createCompany(name: String) {
        
        let company = QSCompany()
        company.name = name
        company.identifier = NSUUID().uuidString
        company.sortIndex.value = companies!.count
        
        realm.beginWrite()
        realm.add(company)
        try! realm.commitWrite()
    }
    
    @IBAction func synchronize() {
        
        synchronize(completion: nil)
    }
    
    override func prepare(for segue: UIStoryboardSegue, sender: Any?) {
        
        if segue.identifier == "showEmployees" {
            
            let indexPath = tableView.indexPathForSelectedRow!
            let company = self.companies![indexPath.row]
            
            let controller = segue.destination as! QSEmployeeTableViewController
            controller.realm = realm
            controller.company = company
            controller.canWrite = true
            controller.navigationItem.leftBarButtonItem = splitViewController?.displayModeButtonItem
            controller.navigationItem.leftItemsSupplementBackButton = true
        }
    }
    
    // MARK : TableView
    
    override func tableView(_ tableView: UITableView, numberOfRowsInSection section: Int) -> Int {
        
        return companies?.count ?? 0
    }
    
    override func tableView(_ tableView: UITableView, cellForRowAt indexPath: IndexPath) -> UITableViewCell {
        
        let cell = tableView.dequeueReusableCell(withIdentifier: "cell") as! CompanyCell
        
        let company = companies![indexPath.row]
        
        cell.nameLabel?.text = company.name
        
        if let _ = synchronizer.share(for: company) {
            cell.sharingButton?.setTitle("Sharing", for: .normal)
        } else {
            cell.sharingButton?.setTitle("Share", for: .normal)
        }
        
        cell.shareButtonAction = { [weak self] in
            self?.shareCompany(company)
        }
        
        return cell
    }
    
    override func tableView(_ tableView: UITableView, canEditRowAt indexPath: IndexPath) -> Bool {
        
        return true
    }
    
    override func tableView(_ tableView: UITableView, commit editingStyle: UITableViewCell.EditingStyle, forRowAt indexPath: IndexPath) {
        
        if editingStyle == .delete {
            
            let company = companies![indexPath.row]
            try! realm.write {
                realm.delete(company)
            }
        }
    }
    
    override func tableView(_ tableView: UITableView, didSelectRowAt indexPath: IndexPath) {
        
        let company = companies![indexPath.row]
        performSegue(withIdentifier: "showEmployees", sender: company)
        
        tableView.deselectRow(at: indexPath, animated: true)
    }
    
    // MARK: - Sync
    
    func synchronize(completion: ((Error?)->())?) {
        
        showLoading(true)
        synchronizer.synchronize { [weak self] (error) in
            
            self?.showLoading(false)
            
            if let error = error {
                
                if (error as NSError).code == CKError.changeTokenExpired.rawValue {
                    self?.appDelegate.didGetChangeTokenExpiredError()
                } else {
                    let alertController = UIAlertController(title: "Error", message: "Error: \(error.localizedDescription)", preferredStyle: .alert)
                    alertController.addAction(UIAlertAction(title: "OK", style: .default, handler: nil))
                    
                    self?.present(alertController, animated: true, completion: nil)
                }
                
            } else {
                
                self?.tableView.reloadData()
                if let zoneID = self?.synchronizer.modelAdapters().first?.recordZoneID() {
                    self?.synchronizer.subscribeForChanges(in: zoneID, completion: { (error) in
                        if let error = error {
                            print("Failed to subscribe with error: \(error.localizedDescription)")
                        } else {
                            print("Subscribed for notifications")
                        }
                    })
                }
            }
            
            completion?(error)
        }
    }
    
    // MARK: - Sharing
    
    func shareCompany(_ company: QSCompany) {
        
        sharingCompany = company
        
        synchronize { [weak self] (error) in
            if error == nil,
                let strongSelf = self {
                
                let sharingController: UICloudSharingController
                let share = strongSelf.synchronizer.share(for: company)
                let container = CKContainer(identifier: strongSelf.synchronizer.containerIdentifier)
                
                if let share = share {
                    share[CKShare.SystemFieldKey.title] = company.name as NSString?
                    sharingController = UICloudSharingController(share: share, container: container)
                } else {
                    sharingController = UICloudSharingController(preparationHandler: { (controller, handler) in
                        
                        strongSelf.synchronizer.share(object: company,
                                                      publicPermission: .readOnly,
                                                      participants: [],
                                                      completion: { (share, error) in
                                                        
                                                        share?[CKShare.SystemFieldKey.title] = company.name as NSString?
                                                        handler(share, container, error)
                        })
                    })
                }
                
                sharingController.availablePermissions = [.allowPublic, .allowReadOnly, .allowReadWrite]
                sharingController.delegate = self
                
                strongSelf.present(sharingController, animated: true, completion: nil)
            }
        }
    }
    
    // MARK : - UICloudSharingControllerDelegate
    
    func itemTitle(for csc: UICloudSharingController) -> String? {
        
        return sharingCompany?.name
    }
    
    func cloudSharingControllerDidSaveShare(_ csc: UICloudSharingController) {
        
        guard let company = sharingCompany,
            let share = csc.share else {
            return
        }
        synchronizer.saveShare(share, for: company)
        tableView.reloadData()
    }
    
    func cloudSharingControllerDidStopSharing(_ csc: UICloudSharingController) {
        
        guard let company = sharingCompany else {
            return
        }
        
        synchronizer.deleteShare(for: company)
        tableView.reloadData()
    }
    
    func cloudSharingController(_ csc: UICloudSharingController, failedToSaveShareWithError error: Error) {
        
        print("Failed to save share: \(error.localizedDescription)")
    }
    
    // MARK: - Loading
    
    func showLoading(_ loading: Bool) {
        
        if loading {
            syncButton?.isHidden = true
            indicatorView?.startAnimating()
        } else {
            syncButton?.isHidden = false
            indicatorView?.stopAnimating()
        }
    }
    
    // MARK: - Resetting
    
    func stopUsingRealmObjects() {
        notificationToken.invalidate()
        sharingCompany = nil
        companies = nil
        tableView.reloadData()
    }
    
}
